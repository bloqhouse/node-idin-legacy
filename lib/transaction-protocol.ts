import { pd } from 'pretty-data'
import { signXML } from './idin-protocol'

import {
  DEFAULT_LANGUAGE,
  EXPIRATION_PERIOD,
  ID_PREFIX,
  KEYNAME,
  LOA,
  MERCHANT_ID,
  MERCHANT_RETURN_URL,
  MERCHANT_SUBID,
  PRIVATE_KEY,
  REQUESTED_SERVICE,
} from './constants'

export function formatTransactionProtocolXML(
  issuerID: string,
  transactionID: string,
  requestedService = REQUESTED_SERVICE,
) {
  const xml = pd.xmlmin(`
    <?xml version="1.0" encoding="UTF-8"?>
    <AcquirerTrxReq version="1.0.0"
      productID="NL:BVN:BankID:1.0"
      xmlns="http://www.betaalvereniging.nl/iDx/messages/Merchant-Acquirer/1.0.0"
      xmlns:ds="http://www.w3.org/2000/09/xmldsig#">
      <createDateTimestamp>${new Date().toISOString()}</createDateTimestamp>
      <Issuer>
        <issuerID>${issuerID}</issuerID>
      </Issuer>
      <Merchant>
        <merchantID>${MERCHANT_ID}</merchantID>
        <subID>0</subID>
        <merchantReturnURL>${MERCHANT_RETURN_URL}</merchantReturnURL>
      </Merchant>
      <Transaction>
        <expirationPeriod>${EXPIRATION_PERIOD}</expirationPeriod>
        <language>${DEFAULT_LANGUAGE}</language>
        <entranceCode>${ID_PREFIX}${transactionID}</entranceCode>
        <container>
          <samlp:AuthnRequest
            xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol"
            xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion"
            AttributeConsumingServiceIndex="${requestedService}"
            ID="${ID_PREFIX}${transactionID}"
            IssueInstant="${new Date().toISOString()}"
            Version="2.0"
            ProtocolBinding="nl:bvn:bankid:1.0:protocol:iDx"
            AssertionConsumerServiceURL="${MERCHANT_RETURN_URL}">
            <saml:Issuer>${MERCHANT_ID}</saml:Issuer>
            <samlp:RequestedAuthnContext Comparison="minimum">
              <saml:AuthnContextClassRef>nl:bvn:bankid:1.0:${LOA}</saml:AuthnContextClassRef>
            </samlp:RequestedAuthnContext>
          </samlp:AuthnRequest>
        </container>
      </Transaction>
    </AcquirerTrxReq>
  `)

  return signXML(xml)
}
