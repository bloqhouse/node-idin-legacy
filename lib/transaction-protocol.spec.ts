import { formatTransactionProtocolXML } from './transaction-protocol'

describe('format status protocol', () => {
  it('it works correctly', () => {
    expect(formatTransactionProtocolXML('XXX', 'XXX')).toMatchSnapshot()
  })
})
