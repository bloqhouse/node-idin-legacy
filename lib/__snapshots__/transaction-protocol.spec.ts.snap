// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`format status protocol it works correctly 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><AcquirerTrxReq version=\\"1.0.0\\" productID=\\"NL:BVN:BankID:1.0\\" xmlns=\\"http://www.betaalvereniging.nl/iDx/messages/Merchant-Acquirer/1.0.0\\" xmlns:ds=\\"http://www.w3.org/2000/09/xmldsig#\\"><createDateTimestamp>2017-10-09T20:08:32.110Z</createDateTimestamp><Issuer><issuerID>XXX</issuerID></Issuer><Merchant><merchantID>0030001645</merchantID><subID>0</subID><merchantReturnURL>http://localhost:8080/txt_return</merchantReturnURL></Merchant><Transaction><expirationPeriod>PT5M</expirationPeriod><language>en</language><entranceCode>BLQXXX</entranceCode><container><samlp:AuthnRequest xmlns:samlp=\\"urn:oasis:names:tc:SAML:2.0:protocol\\" xmlns:saml=\\"urn:oasis:names:tc:SAML:2.0:assertion\\" AttributeConsumingServiceIndex=\\"21968\\" ID=\\"BLQXXX\\" IssueInstant=\\"2017-10-09T20:08:32.110Z\\" Version=\\"2.0\\" ProtocolBinding=\\"nl:bvn:bankid:1.0:protocol:iDx\\" AssertionConsumerServiceURL=\\"http://localhost:8080/txt_return\\"><saml:Issuer>0030001645</saml:Issuer><samlp:RequestedAuthnContext Comparison=\\"minimum\\"><saml:AuthnContextClassRef>nl:bvn:bankid:1.0:loa2</saml:AuthnContextClassRef></samlp:RequestedAuthnContext></samlp:AuthnRequest></container></Transaction><Signature xmlns=\\"http://www.w3.org/2000/09/xmldsig#\\"><SignedInfo><CanonicalizationMethod Algorithm=\\"http://www.w3.org/2001/10/xml-exc-c14n#\\"/><SignatureMethod Algorithm=\\"http://www.w3.org/2001/04/xmldsig-more#rsa-sha256\\"/><Reference URI=\\"\\"><Transforms><Transform Algorithm=\\"http://www.w3.org/2000/09/xmldsig#enveloped-signature\\"/><Transform Algorithm=\\"http://www.w3.org/2001/10/xml-exc-c14n#\\"/></Transforms><DigestMethod Algorithm=\\"http://www.w3.org/2001/04/xmlenc#sha256\\"/><DigestValue>2Zsl05p7ZvWaZFn5hSDuE4OYjhEcL7ySYEkDnlt6ap4=</DigestValue></Reference></SignedInfo><SignatureValue>kH9Thyho0cnwDWd2/b+Vnh7rvETDvUIOw207ftTSIOR7zmTCRe2Qv9TQq6raTZqfzl+AVbC0rDSGXwTc9Df3599lCByK6LbwcnVPJy0s1PRPhcnsc001O21PhET/XKajnLQltjcNkSOA+hupIwO5+wXOOn5YwTrTNg9eE/Ktz5cs5S0fSqW1O/hQvx6IKGNDXkyNYQKp5K52pb5T6cxBrqP3/lRTTMY6RpdjyM1jHdzFT9QxLA7lBgFceX6FVH9kRHmrfMKrxhnPCdcREGq2pq3HLvEB0jEv4PwbRkqwuBYy6k/GF4F6F5VAhBfotIF/NTbzCrwxwkfYgKMHjO6k4w==</SignatureValue><KeyInfo><KeyName>8c6415e72e553d2a9bda7afc9fcc0a5ec5d82841</KeyName></KeyInfo></Signature></AcquirerTrxReq>"`;
